src/peft/tuners/tuners_utils.py

```diff
@@ -517,18 +540,66 @@ def inject_adapter(
             if excluded_modules and excluded_modules[-1] == key:
                 continue
 
-            result = self._check_target_module_exists(peft_config, key)
-            if isinstance(result, _ExcludedModule):
-                excluded_modules.append(key)
-            elif not result:
-                unmatched_modules.append(key)
+            if state_dict is None:
+                # normal mechanism: match the modules using the peft_config
+                result = self._check_target_module_exists(peft_config, key)
+                if isinstance(result, _ExcludedModule):
+                    excluded_modules.append(key)
+                elif not result:
+                    unmatched_modules.append(key)
+                else:
+                    self.targeted_module_names.append(key)
+                    parent, target, target_name = _get_submodules(model, key)
+                    self._check_target_module_compatiblity(peft_config, model, target_name)
+                    ctx = init_empty_weights if low_cpu_mem_usage else nullcontext
+                    with ctx():
+                        self._create_and_replace(
+                            peft_config, adapter_name, target, target_name, parent, current_key=key
+                        )
             else:
-                self.targeted_module_names.append(key)
-                parent, target, target_name = _get_submodules(model, key)
-                self._check_target_module_compatiblity(peft_config, model, target_name)
-                ctx = init_empty_weights if low_cpu_mem_usage else nullcontext
-                with ctx():
-                    self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)
+                # use the state_dict to match modules instead
+                if key not in state_dict_keys:
```

Comment:
In case, `state_dict` is none these keys will be empty. Just flagging.

Reply:
We check above if `state_dict` is `None` or not, so this is safe here.